### ISSUE TITLE
Get rid of clearVariables and allow object reuse

### ISSUE THREAD

**Issue by felipeochoa** [dolanmiu/docx#13]
So I think I understand `clearVariable` and the `Formatter` much better now, and have an idea for how to proceed. Let me first lay out what I think these are doing and why to make sure what I'm suggesting makes sense.

1. Ultimately, the point of the `Formatter` is to convert the `docx` objects into something that `xml`, with its wonky format, can understand.
2. The `Packer` is the only entity responsible for calling the `Formatter`, and the `Formatter` in turn calls `clearVariables`
3. The reason to have this separate formatting step is that `xml`'s format is otherwise unwieldy to use as the structures are built up.

The current approach works, but has a few shortcomings:

* Existing objects are rendered unstable after `clearVariables` has been called on them. This is causing me particular trouble, since it means I have to re-create all styles, numbering and properties from scratch for every document I render
* All classes that use custom fields must define `clearVariables` and take care to call their children's `clearVariables`. 
    + This makes it tricky to define new classes/fields, and can easily lead to bugs that won't be caught until cross-checked with Word
    + The variable clearing logic is scattered all over the place
* The formatter performs a number of steps, including `JSON.parse(JSON.stringify(...))`, and has to traverse the entire structure (slow and memory intensive)

----------------------------------------------------------------------------------------------------------

If you agree with me this far, let me know what you think of the following. Right now, there are basically three families of components, all inheriting from `BaseXml`:

* `XmlComponent`: This is the most common parent class, and it serializes to an XML tag
* `XmlAttributeComponent`: This is also quite common, it serializes to the attributes on a tag. It's thus commong for `XmlComponent` classes to have an associated `XmlAttributeComponent`
* `XmlUnit`: This class is another way of representing attributes on a tag. Unlike `XmlAttributeComponent`, Each instance represents one attribute. It's currently only used for the top-level properties (Title, Subject, etc.)

I'm working on a patch that will:

* Combine `XmlUnit` and `XmlComponent` into one class that can take either strings or child components as children

* Make `XmlComponent` into a generic class, parametrized by the class of attributes it uses, which are now always stored in its `.attrs` field. E.g., `Paragraph` would now be defined as:

   ```ts
   export class Paragraph extends XmlComponent<ParagraphProperties> { ... }
   ```

* Simplify the `Formatter`, using the fact that all `BaseXml` instances are either an `XmlComponent` or one of their `.attrs`. This is the magic step, and actually becomes quite easy once the other two are done. Basically, instead of mutating the `XmlComponent` into the `xml` format, the `Formatter` will create a new object using the data from the component but a brand top-level structure. Since we're already round-tripping through JSON, this new approach will be more efficient, since there will be structures we can re-use (e.g., the array of child elements, `.root`). And it also avoids having to remove extraneous fields from `XmlComponent`

I should have a PR in the next hour or so, but wanted to give you a heads up of what was coming!

**Comment by dolanmiu** [comment]
Yes you are right with clearVariables(). Could you find a class with has an implementation in clearVariables()?

Yes I think deprecating XmlUnit; and combining it into one class is a good idea.

Wow yes, I like the idea of using XmlComponent<Properties>, will be much cleaner.

Regarding the JSON.parse thing, At the time it was quick and understandable to me. Will the new implantation of the formatter be easy to read? I don't mind sacrificing some efficiency for readability

**Comment by felipeochoa** [comment]
Almost there with the first part, but yes, I think it will be much easier to understand actually. Something like:

```js
formatComponent(comp) {
  const ret = {};
  ret[comp.rootKey] = comp.root.map.forEach(subComp => {
    if (subComp instanceof XmlComponent)
      return formatComponent(comp);
    else  // Must be a string
      return comp
  });
  return ret;
}
```

(Or something like that -- I just wrote that quickly in the ocmment box)

**Comment by dolanmiu** [comment]
is it possible to have the format like the clearVariables() kind of style? I say this because then the format logic for each component can be inside the component itself, which could make it very clean. Just a thought.

Also, we should use native for-of rather than lodashes version (foreach)

**Comment by felipeochoa** [comment]
I like that idea! How about giving elements a `.toXml` method? We'd only have to implement it for `XmlComponent` and `XmlAttributeComponent`

**Comment by dolanmiu** [comment]
Yeah a .toXml() method is perfect. Nice and descriptive

**Comment by dolanmiu** [comment]
I am looking into the **tests** now, there is the `jsonify()` method which does `JSON.parse/JSON.stringify`

I am moving that out into a Utility class because its copied and pasted everywhere.

This got me thinking, surely we would need `JSON.parse/JSON.stringify` in the formatter as well then?

**Comment by dolanmiu** [comment]
Wow, I just saw you did this:

```ts
expect(tree).to.deep.equal({
     "w:spacing": [{_attr: {"w:before": 100}}],
});
```

which eliminates the need for the `jsonify()` method? Epic stuff

**Comment by felipeochoa** [comment]
Closed with #14
